### PR TITLE
[21.09] Handle unique tag name constraints

### PR DESCRIPTION
**Pull request by mvdbeek**
The idea is to create the tag one by one in a separate session, and
catch any violations.

Fixes https://github.com/galaxyproject/galaxy/issues/13213

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

File: lib/galaxy/model/tags.py [hunk 1]
```diff
@@ -245,7 +247,18 @@ def _get_tag(self, tag_name):
...
     def _create_tag_instance(self, tag_name):
-        return galaxy.model.Tag(type=0, name=tag_name)
+        # For good performance caller should first check if there's already an appropriate tag
+        Session = sessionmaker(self.sa_session.bind)
+        tag = galaxy.model.Tag(type=0, name=tag_name)
+        with Session() as separate_session:
+            separate_session.add(tag)
+            try:
+                separate_session.commit()
+                separate_session.flush()
```
Comment (jdavcs) [inline lib/galaxy/model/tags.py:257]:
```suggestion
```
already flushed via `commit()`

**Review by jdavcs — APPROVED** [review]
This should work.

File: lib/galaxy/model/tags.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/model/tags.py:257]: That was my intuition too, but the unit tests fail without this. I haven't checked out why, but my guess is that the sessionfactory parameters aren't picked up. (and maybe that explains why our similar attempts in `ToolShedRepositoryCache` were doomed to fail ...)

Comment (jdavcs) [reply lib/galaxy/model/tags.py:257]: My local unit tests are not failing. What is the test that's failing on your setup?

I see that the values of these params `autoflush=False, autocommit=True` are reversed on the created session (i.e., `self.sa_session` vs `separate_session`). I'll try to figure out how to fix this.

Comment (jdavcs) [reply lib/galaxy/model/tags.py:257]: OK, so getting the session factory params to apply is easy: `Session.configure(autoflush=False, autocommit=True)` is the immediate "fix". Of course, a better way to do it is to have a function somewhere in `model.base` (or similar location) that can be called anytime we need a one-time separate, properly configured session object. Could be a context manager for good measure. 

However, right now that will raise an error (maybe that's the error you were seeing). The problem is the autocommit mode. As per docs:  "When using legacy “autocommit” mode, this method is only valid to call if a transaction is actually in progress, else an error is raised." ([ref](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.commit))

I think what we have here is sufficient for 21.09. Once we address the autocommit setting (as part of the move to SA 2.0), we can revisit this, etc.

**Pull request merged**